### ISSUE TITLE
New Linux mount API

### ISSUE THREAD

**Issue by mahkoh** [nix-rust/nix#1237]
I would like to add support for the new Linux mount API. This API currently does not have man pages but I'll attach some recent drafts below. Would such a Linux-specific API be suitable for this project?

**Comment by mahkoh** [comment]
NAME
====

fsopen, fsmount - Filesystem parameterisation and mount creation

SYNOPSIS
========

    #include <sys/types.h>


    #include <sys/mount.h>


    #include <unistd.h>


    #include <fcntl.h> /* Definition of AT_* constants */


    #include <sys/mount.h> 

    int fsopen(const char *fsname, unsigned int flags);

    int fsmount(int fd, unsigned int flags, unsigned int mount_attrs);

*Note*: There are no glibc wrappers for these system calls.

DESCRIPTION
===========

**fsopen**() creates a blank filesystem configuration context within the
kernel for the filesystem named in the *fsname* parameter, puts it into
creation mode and attaches it to a file descriptor, which it then
returns. The file descriptor can be marked close-on-exec by setting
**FSOPEN\_CLOEXEC** in *flags*.

After calling fsopen(), the file descriptor should be passed to the
**fsconfig**(2) system call, using that to specify the desired
filesystem and security parameters.

When the parameters are all set, the **fsconfig**() system call should
then be called again with **FSCONFIG\_CMD\_CREATE** as the command
argument to effect the creation.

> **\[!\] NOTE**: Depending on the filesystem type and parameters, this
> may rather share an existing in-kernel filesystem representation
> instead of creating a new one. In such a case, the parameters
> specified may be discarded or may overwrite the parameters set by a
> previous mount - at the filesystem\'s discretion.

The file descriptor also serves as a channel by which more comprehensive
error, warning and information messages may be retrieved from the kernel
using **read**(2).

Once the creation command has been successfully run on a context, the
context is switched into need-mount mode which prevents further
configuration. At this point, **fsmount**() should be called to create a
mount object.

**fsmount**() takes the file descriptor returned by **fsopen**() and
creates a mount object for the filesystem root specified there. The
attributes of the mount object are set from the *mount\_attrs*
parameter. The attributes specify the propagation and mount restrictions
to be applied to accesses through this mount.

The mount object is then attached to a new file descriptor that looks
like one created by **open**(2) with **O\_PATH** or **open\_tree**(2).
This can be passed to **move\_mount**(2) to attach the mount object to a
mountpoint, thereby completing the process.

The file descriptor returned by fsmount() is marked close-on-exec if
FSMOUNT\_CLOEXEC is specified in *flags*.

After fsmount() has completed, the context created by fsopen() is reset
and moved to reconfiguration state, allowing the new superblock to be
reconfigured. See **fspick**(2) for details.

Message Retrieval Interface
---------------------------

The context file descriptor may be queried for message strings at any
time by calling **read**(2) on the file descriptor. This will return
formatted messages that are prefixed to indicate their class:

**\"e \<message\>\"**

:   An error message string was logged.

**\"i \<message\>\"**

:   An informational message string was logged.

**\"w \<message\>\"**

:   An warning message string was logged.

Messages are removed from the queue as they\'re read.

EXAMPLES
========

To illustrate the process, here\'s an example whereby this can be used
to mount an ext4 filesystem on /dev/sdb1 onto /mnt.

    sfd = fsopen("ext4", FSOPEN_CLOEXEC);
    fsconfig(sfd, FSCONFIG_SET_FLAG, "ro", NULL, 0);
    fsconfig(sfd, FSCONFIG_SET_STRING, "source", "/dev/sdb1", 0);
    fsconfig(sfd, FSCONFIG_SET_FLAG, "noatime", NULL, 0);
    fsconfig(sfd, FSCONFIG_SET_FLAG, "acl", NULL, 0);
    fsconfig(sfd, FSCONFIG_SET_FLAG, "user_attr", NULL, 0);
    fsconfig(sfd, FSCONFIG_SET_FLAG, "iversion", NULL, 0);
    fsconfig(sfd, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
    mfd = fsmount(sfd, FSMOUNT_CLOEXEC, MS_RELATIME);
    move_mount(mfd, "", sfd, AT_FDCWD, "/mnt", MOVE_MOUNT_F_EMPTY_PATH);

Here, an ext4 context is created first and attached to sfd. This is then
told where its source will be, given a bunch of options and created.
Then fsmount() is called to create a mount object and **move\_mount**(2)
is called to attach it to its intended mountpoint.

And here\'s an example of mounting from an NFS server and setting a
Smack security module label on it too:

    sfd = fsopen("nfs", 0);
    fsconfig(sfd, FSCONFIG_SET_STRING, "source", "example.com/pub/linux", 0);
    fsconfig(sfd, FSCONFIG_SET_STRING, "nfsvers", "3", 0);
    fsconfig(sfd, FSCONFIG_SET_STRING, "rsize", "65536", 0);
    fsconfig(sfd, FSCONFIG_SET_STRING, "wsize", "65536", 0);
    fsconfig(sfd, FSCONFIG_SET_STRING, "smackfsdef", "foolabel", 0);
    fsconfig(sfd, FSCONFIG_SET_FLAG, "rdma", NULL, 0);
    fsconfig(sfd, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
    mfd = fsmount(sfd, 0, MS_NODEV);
    move_mount(mfd, "", sfd, AT_FDCWD, "/mnt", MOVE_MOUNT_F_EMPTY_PATH);

RETURN VALUE
============

On success, both functions return a file descriptor. On error, -1 is
returned, and *errno* is set appropriately.

ERRORS
======

The error values given below result from filesystem type independent
errors. Each filesystem type may have its own special errors and its own
special behavior. See the Linux kernel source code for details.

**EBUSY**

:   The context referred to by *fd* is not in the right state to be used
    by **fsmount**().

**EFAULT**

:   One of the pointer arguments points outside the user address space.

**EINVAL**

:   *flags* had an invalid flag set.

**EINVAL**

:   *mount\_attrs,* includes invalid **MOUNT\_ATTR\_\*** flags.

**EMFILE**

:   The system has too many open files to create more.

**ENFILE**

:   The process has too many open files to create more.

**ENODEV**

:   Filesystem *fsname* not configured in the kernel.

**ENOMEM**

:   The kernel could not allocate sufficient memory to complete the
    call.

**EPERM**

:   The caller does not have the required privileges.

CONFORMING TO
=============

These functions are Linux-specific and should not be used in programs
intended to be portable.

VERSIONS
========

**fsopen**(), and **fsmount**() were added to Linux in kernel 5.1.

NOTES
=====

Glibc does not (yet) provide a wrapper for the **fsopen**() or
**fsmount**() system calls; call them using **syscall**(2).

SEE ALSO
========

**mountpoint**(1), **fsconfig**(2), **fspick**(2), **move\_mount**(2),
**open\_tree**(2), **umount**(2), **mount\_namespaces**(7),
**path\_resolution**(7), **findmnt**(8), **lsblk**(8), **mount**(8),
**umount**(8)

**Comment by mahkoh** [comment]
NAME
====

fsconfig - Filesystem parameterisation

SYNOPSIS
========

    #include <sys/types.h>


    #include <sys/mount.h>


    #include <unistd.h>


    #include <sys/mount.h>

    int fsconfig(int *fd, unsigned int cmd, const char *key,


     const void __user *value, int aux);

*Note*: There is no glibc wrapper for this system call.

DESCRIPTION
===========

**fsconfig**() is used to supply parameters to and issue commands
against a filesystem configuration context as set up by **fsopen**(2) or
**fspick**(2). The context is supplied attached to the file descriptor
specified by *fd* argument.

The *cmd* argument indicates the command to be issued, where some of the
commands simply supply parameters to the context. The meaning of *key*,
*value* and *aux* are command-dependent; unless required for the
command, these should be set to NULL or 0.

The available commands are:

**FSCONFIG\_SET\_FLAG**

:   Set the parameter named by *key* to true. This may incur error
    **EINVAL** if the parameter requires an argument.

**FSCONFIG\_SET\_STRING**

:   Set the parameter named by *key* to a string. This may incur error
    **EINVAL** if the parser doesn\'t want a parameter here, wants a
    non-string or the string cannot be interpreted appropriately.
    *value* points to a NUL-terminated string.

**FSCONFIG\_SET\_BINARY**

:   Set the parameter named by *key* to be a binary blob argument. This
    may cause **EINVAL** to be returned if the filesystem parser isn\'t
    expecting a binary blob and it can\'t be converted to something
    usable. *value* points to the data and *aux* indicates the size of
    the data.

**FSCONFIG\_SET\_PATH**

:   Set the parameter named by *key* to the object at the provided path.
    *value* should point to a NULL-terminated pathname string and aux
    may indicate **AT\_FDCWD** or a file descriptor indicating a
    directory from which to begin a relative pathwalk. This may return
    any errors incurred by the pathwalk and may return **EINVAL** if the
    parameter isn\'t expecting a path.

    Note that FSCONFIG\_SET\_STRING can be used instead, implying
    AT\_FDCWD.

**FSCONFIG\_SET\_PATH\_EMPTY**

:   As FSCONFIG\_SET\_PATH, but with **AT\_EMPTY\_PATH** applied to the
    pathwalk.

**FSCONFIG\_SET\_FD**

:   Set the parameter named by *key* to the file descriptor specified by
    *aux*. This will incur **EINVAL** if the parameter doesn\'t expect a
    file descriptor or **EBADF** if the file descriptor is invalid.

    Note that FSCONFIG\_SET\_STRING can be used instead with the file
    descriptor passed as a decimal string.

**FSCONFIG\_CMD\_CREATE**

:   This command causes the filesystem to take the parameters set in the
    context and to try to create filesystem representation in the
    kernel. If it can share an existing one, it may do that instead if
    the filesystem type and parameters permit that. This is intended for
    use with **fsopen**(2).

**FSCONFIG\_CMD\_RECONFIGURE**

:   This command causes the filesystem to apply the parameters set in
    the context to an already existing filesystem representation in
    memory and to alter it. This is intended for use with **fspick**(2),
    but may also by used against the context created by **fsopen()**
    after **fsmount**(2) has been called on it.

EXAMPLES
========

    fsconfig(sfd, FSCONFIG_SET_FLAG, "ro", NULL, 0);

    fsconfig(sfd, FSCONFIG_SET_STRING, "user_xattr", "false", 0);

    fsconfig(sfd, FSCONFIG_SET_BINARY, "ms_pac", pac_buffer, pac_size);

    fsconfig(sfd, FSCONFIG_SET_PATH, "journal", "/dev/sdd4", AT_FDCWD);

    dirfd = open("/dev/", O_PATH);
    fsconfig(sfd, FSCONFIG_SET_PATH, "journal", "sdd4", dirfd);

    fd = open("/overlays/mine/", O_PATH);
    fsconfig(sfd, FSCONFIG_SET_PATH_EMPTY, "lower_dir", "", fd);

    pipe(pipefds);
    fsconfig(sfd, FSCONFIG_SET_FD, "fd", NULL, pipefds[1]);

RETURN VALUE
============

On success, the function returns 0. On error, -1 is returned, and
*errno* is set appropriately.

ERRORS
======

The error values given below result from filesystem type independent
errors. Each filesystem type may have its own special errors and its own
special behavior. See the Linux kernel source code for details.

**EACCES**

:   A component of a path was not searchable. (See also
    **path\_resolution**(7).)

**EACCES**

:   Mounting a read-only filesystem was attempted without specifying the
    \'**ro**\' parameter.

**EACCES**

:   A specified block device is located on a filesystem mounted with the
    **MS\_NODEV** option.

**EBADF**

:   The file descriptor given by *fd* or possibly by *aux* (depending on
    the command) is invalid.

**EBUSY**

:   The context attached to *fd* is in the wrong state for the given
    command.

**EBUSY**

:   The filesystem representation cannot be reconfigured read-only
    because it still holds files open for writing.

**EFAULT**

:   One of the pointer arguments points outside the user address space.

**EINVAL**

:   *fd* does not refer to a filesystem configuration context.

**EINVAL**

:   One of the source parameters referred to an invalid superblock.

**ELOOP**

:   Too many links encountered during pathname resolution.

**ENAMETOOLONG**

:   A path name was longer than **MAXPATHLEN**.

**ENOENT**

:   A pathname was empty or had a nonexistent component.

**ENOMEM**

:   The kernel could not allocate sufficient memory to complete the
    call.

**ENOTBLK**

:   Once of the parameters does not refer to a block device (and a
    device was required).

**ENOTDIR**

:   *pathname*, or a prefix of *source*, is not a directory.

**EOPNOTSUPP**

:   The command given by *cmd* was not valid.

**ENXIO**

:   The major number of a block device parameter is out of range.

**EPERM**

:   The caller does not have the required privileges.

CONFORMING TO
=============

These functions are Linux-specific and should not be used in programs
intended to be portable.

VERSIONS
========

**fsconfig**() was added to Linux in kernel 5.1.

NOTES
=====

Glibc does not (yet) provide a wrapper for the **fspick**() system call;
call it using **syscall**(2).

SEE ALSO
========

**mountpoint**(1), **fsmount**(2), **fsopen**(2), **fspick**(2),
**mount\_namespaces**(7), **path\_resolution**(7)

**Comment by mahkoh** [comment]
NAME
====

fspick - Select filesystem for reconfiguration

SYNOPSIS
========

    #include <sys/types.h>


    #include <sys/mount.h>


    #include <unistd.h>


    #include <fcntl.h> /* Definition of AT_* constants */

    int fspick(int dirfd, const char *pathname, unsigned int flags);

*Note*: There is no glibc wrapper for this system call.

DESCRIPTION
===========

**fspick**() creates a new filesystem configuration context within the
kernel and attaches a pre-existing superblock to it so that it can be
reconfigured (similar to **mount**(8) with the \"-o remount\" option).
The configuration context is marked as being in reconfiguration mode and
attached to a file descriptor, which is returned to the caller. This can
be marked close-on-exec by setting **FSPICK\_CLOEXEC** in *flags*.

The target is whichever superblock backs the object determined by *dfd*,
*pathname* and *flags*. The following can be set in *flags* to control
the pathwalk to that object:

**FSPICK\_SYMLINK\_NOFOLLOW**

:   Don\'t follow symbolic links in the terminal component of the path.

**FSPICK\_NO\_AUTOMOUNT**

:   Don\'t follow automounts in the terminal component of the path.

**FSPICK\_EMPTY\_PATH**

:   Allow an empty string to be specified as the pathname. This allows
    *dirfd* to specify a path exactly.

After calling fspick(), the file descriptor should be passed to the
**fsconfig**(2) system call, using that to specify the desired changes
to filesystem and security parameters.

When the parameters are all set, the **fsconfig**() system call should
then be called again with **FSCONFIG\_CMD\_RECONFIGURE** as the command
argument to effect the reconfiguration.

After the reconfiguration has taken place, the context is wiped clean
(apart from the superblock attachment, which remains) and can be reused
to make another reconfiguration.

The file descriptor also serves as a channel by which more comprehensive
error, warning and information messages may be retrieved from the kernel
using **read**(2).

Message Retrieval Interface
---------------------------

The context file descriptor may be queried for message strings at any
time by calling **read**(2) on the file descriptor. This will return
formatted messages that are prefixed to indicate their class:

**\"e \<message\>\"**

:   An error message string was logged.

**\"i \<message\>\"**

:   An informational message string was logged.

**\"w \<message\>\"**

:   An warning message string was logged.

Messages are removed from the queue as they\'re read and the queue has a
limited depth, so it\'s possible for some to get lost.

EXAMPLES
========

To illustrate the process, here\'s an example whereby this can be used
to reconfigure a filesystem:

    sfd = fspick(AT_FDCWD, "/mnt", FSPICK_NO_AUTOMOUNT | FSPICK_CLOEXEC);
    fsconfig(sfd, FSCONFIG_SET_FLAG, "ro", NULL, 0);
    fsconfig(sfd, FSCONFIG_SET_STRING, "user_xattr", "false", 0);
    fsconfig(sfd, FSCONFIG_CMD_RECONFIGURE, NULL, NULL, 0);

RETURN VALUE
============

On success, the function returns a file descriptor. On error, -1 is
returned, and *errno* is set appropriately.

ERRORS
======

The error values given below result from filesystem type independent
errors. Each filesystem type may have its own special errors and its own
special behavior. See the Linux kernel source code for details.

**EACCES**

:   A component of a path was not searchable. (See also
    **path\_resolution**(7).)

**EFAULT**

:   *pathname* points outside the user address space.

**EINVAL**

:   *flags* includes an undefined value.

**ELOOP**

:   Too many links encountered during pathname resolution.

**EMFILE**

:   The system has too many open files to create more.

**ENFILE**

:   The process has too many open files to create more.

**ENAMETOOLONG**

:   A pathname was longer than **MAXPATHLEN**.

**ENOENT**

:   A pathname was empty or had a nonexistent component.

**ENOMEM**

:   The kernel could not allocate sufficient memory to complete the
    call.

**EPERM**

:   The caller does not have the required privileges.

CONFORMING TO
=============

These functions are Linux-specific and should not be used in programs
intended to be portable.

VERSIONS
========

**fsopen**(), **fsmount**() and **fspick**() were added to Linux in
kernel 5.1.

NOTES
=====

Glibc does not (yet) provide a wrapper for the **fspick**() system call;
call it using **syscall**(2).

SEE ALSO
========

**mountpoint**(1), **fsconfig**(2), **fsopen**(2),
**path\_resolution**(7), **mount**(8)

**Comment by mahkoh** [comment]
NAME
====

move\_mount - Move mount objects around the filesystem topology

SYNOPSIS
========

    #include <sys/types.h>


    #include <sys/mount.h>


    #include <unistd.h>


    #include <fcntl.h> /* Definition of AT_* constants */

    int move_mount(int from_dirfd, const char *from_pathname,
     int to_dirfd, const char *to_pathname,
     unsigned int flags);

*Note*: There is no glibc wrapper for this system call.

DESCRIPTION
===========

The **move\_mount**() call moves a mount from one place to another; it
can also be used to attach an unattached mount created by **fsmount**()
or **open\_tree**() with **OPEN\_TREE\_CLONE**.

If **move\_mount**() is called repeatedly with a file descriptor that
refers to a mount object, then the object will be attached/moved the
first time and then moved again and again and again, detaching it from
the previous mountpoint each time.

To access the source mount object or the destination mountpoint, no
permissions are required on the object itself, but if either pathname is
supplied, execute (search) permission is required on all of the
directories specified in *from\_pathname* or *to\_pathname*.

The caller does, however, require the appropriate capabilities or
permission to effect a mount.

**move\_mount**() uses *from\_pathname*, *from\_dirfd* and part of
*flags* to locate the mount object to be moved and *to\_pathname*,
*to\_dirfd* and another part of *flags* to locate the destination
mountpoint. Each lookup can be done in one of a variety of ways:

\[\*\] By absolute path.

:   The pathname points to an absolute path and the dirfd is ignored.
    The file is looked up by name, starting from the root of the
    filesystem as seen by the calling process.

\[\*\] By cwd-relative path.

:   The pathname points to a relative path and the dirfd is *AT\_FDCWD*.
    The file is looked up by name, starting from the current working
    directory.

\[\*\] By dir-relative path.

:   The pathname points to relative path and the dirfd indicates a file
    descriptor pointing to a directory. The file is looked up by name,
    starting from the directory specified by *dirfd*.

\[\*\] By file descriptor.

:   The pathname points to \"\", the dirfd points directly to the mount
    object to move or the destination mount point and the appropriate
    **\*\_EMPTY\_PATH** flag is set.

*flags* can be used to influence a path-based lookup. A value for
*flags* is constructed by OR\'ing together zero or more of the following
constants:

**MOVE\_MOUNT\_F\_EMPTY\_PATH**

:   If *from\_pathname* is an empty string, operate on the file referred
    to by *from\_dirfd* (which may have been obtained using the
    **open**(2) **O\_PATH** flag or **open\_tree**()) If *from\_dirfd*
    is **AT\_FDCWD**, the call operates on the current working
    directory. In this case, *from\_dirfd* can refer to any type of
    file, not just a directory. This flag is Linux-specific; define
    **\_GNU\_SOURCE** to obtain its definition.

**MOVE\_MOUNT\_T\_EMPTY\_PATH**

:   As above, but operating on *to\_pathname* and *to\_dirfd*.

**MOVE\_MOUNT\_F\_NO\_AUTOMOUNT**

:   Don\'t automount the terminal (\"basename\") component of
    *from\_pathname* if it is a directory that is an automount point.
    This allows a mount object that has an automount point at its root
    to be moved and prevents unintended triggering of an automount
    point. The **MOVE\_MOUNT\_F\_NO\_AUTOMOUNT** flag has no effect if
    the automount point has already been mounted over. This flag is
    Linux-specific; define **\_GNU\_SOURCE** to obtain its definition.

**MOVE\_MOUNT\_T\_NO\_AUTOMOUNT**

:   As above, but operating on *to\_pathname* and *to\_dirfd*. This
    allows an automount point to be manually mounted over.

**MOVE\_MOUNT\_F\_SYMLINKS**

:   If *from\_pathname* is a symbolic link, then dereference it. The
    default for **move\_mount**() is to not follow symlinks.

**MOVE\_MOUNT\_T\_SYMLINKS**

:   As above, but operating on *to\_pathname* and *to\_dirfd*.

EXAMPLES
========

The **move\_mount**() function can be used like the following:

>     move_mount(AT_FDCWD, "/a", AT_FDCWD, "/b", 0);

This would move the object mounted on \"/a\" to \"/b\". It can also be
used in conjunction with **open\_tree**(2) or **open**(2) with
**O\_PATH**:

>     fd = open_tree(AT_FDCWD, "/mnt", 0);
>     move_mount(fd, "", AT_FDCWD, "/mnt2", MOVE_MOUNT_F_EMPTY_PATH);
>     move_mount(fd, "", AT_FDCWD, "/mnt3", MOVE_MOUNT_F_EMPTY_PATH);
>     move_mount(fd, "", AT_FDCWD, "/mnt4", MOVE_MOUNT_F_EMPTY_PATH);

This would attach the path point for \"/mnt\" to fd, then it would move
the mount to \"/mnt2\", then move it to \"/mnt3\" and finally to
\"/mnt4\".

It can also be used to attach new mounts:

>     sfd = fsopen("ext4", FSOPEN_CLOEXEC);
>     fsconfig(sfd, FSCONFIG_SET_STRING, "source", "/dev/sda1", 0);
>     fsconfig(sfd, FSCONFIG_SET_FLAG, "user_xattr", NULL, 0);
>     fsconfig(sfd, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
>     mfd = fsmount(sfd, FSMOUNT_CLOEXEC, MOUNT_ATTR_NODEV);
>     move_mount(mfd, "", AT_FDCWD, "/home", MOVE_MOUNT_F_EMPTY_PATH);

Which would open the Ext4 filesystem mounted on \"/dev/sda1\", turn on
user extended attribute support and create a mount object for it.
Finally, the new mount object would be attached with **move\_mount**()
to \"/home\".

RETURN VALUE
============

On success, 0 is returned. On error, -1 is returned, and *errno* is set
appropriately.

ERRORS
======

**EACCES**

:   Search permission is denied for one of the directories in the path
    prefix of *pathname*. (See also **path\_resolution**(7).)

**EBADF**

:   *from\_dirfd* or *to\_dirfd* is not a valid open file descriptor.

**EFAULT**

:   *from\_pathname* or *to\_pathname* is NULL or either one point to a
    location outside the process\'s accessible address space.

**EINVAL**

:   Reserved flag specified in *flags*.

**ELOOP**

:   Too many symbolic links encountered while traversing the pathname.

**ENAMETOOLONG**

:   *from\_pathname* or *to\_pathname* is too long.

**ENOENT**

:   A component of *from\_pathname* or *to\_pathname* does not exist, or
    one is an empty string and the appropriate **\*\_EMPTY\_PATH** was
    not specified in *flags*.

**ENOMEM**

:   Out of memory (i.e., kernel memory).

**ENOTDIR**

:   A component of the path prefix of *from\_pathname* or *to\_pathname*
    is not a directory or one or the other is relative and the
    appropriate *\*\_dirfd* is a file descriptor referring to a file
    other than a directory.

VERSIONS
========

**move\_mount**() was added to Linux in kernel 4.18.

CONFORMING TO
=============

**move\_mount**() is Linux-specific.

NOTES
=====

Glibc does not (yet) provide a wrapper for the **move\_mount**() system
call; call it using **syscall**(2).

SEE ALSO
========

**fsmount**(2), **fsopen**(2), **open\_tree**(2)

**Comment by mahkoh** [comment]
NAME
====

open\_tree - Pick or clone mount object and attach to fd

SYNOPSIS
========

    #include <sys/types.h>


    #include <sys/mount.h>


    #include <unistd.h>


    #include <fcntl.h> /* Definition of AT_* constants */

    int open_tree(int dirfd, const char *pathname, unsigned int flags);

*Note*: There are no glibc wrappers for these system calls.

DESCRIPTION
===========

**open\_tree**() picks the mount object specified by the pathname and
attaches it to a new file descriptor or clones it and attaches the clone
to the file descriptor. The resultant file descriptor is
indistinguishable from one produced by **open**(2) with **O\_PATH**.

In the case that the mount object is cloned, the clone will be
\"unmounted\" and destroyed when the file descriptor is closed if it is
not otherwise mounted somewhere by calling **move\_mount**(2).

To select a mount object, no permissions are required on the object
referred to by the path, but execute (search) permission is required on
all of the directories in *pathname* that lead to the object.

To clone an object, however, the caller must have mount capabilities and
permissions.

**open\_tree**() uses *pathname*, *dirfd* and *flags* to locate the
target object in one of a variety of ways:

\[\*\] By absolute path.

:   *pathname* points to an absolute path and *dirfd* is ignored. The
    object is looked up by name, starting from the root of the
    filesystem as seen by the calling process.

\[\*\] By cwd-relative path.

:   *pathname* points to a relative path and *dirfd* is *AT\_FDCWD*. The
    object is looked up by name, starting from the current working
    directory.

\[\*\] By dir-relative path.

:   *pathname* points to relative path and *dirfd* indicates a file
    descriptor pointing to a directory. The object is looked up by name,
    starting from the directory specified by *dirfd*.

\[\*\] By file descriptor.

:   *pathname* is \"\", *dirfd* indicates a file descriptor and
    **AT\_EMPTY\_PATH** is set in *flags*. The mount attached to the
    file descriptor is queried directly. The file descriptor may point
    to any type of file, not just a directory.

*flags* can be used to control the operation of the function and to
influence a path-based lookup. A value for *flags* is constructed by
OR\'ing together zero or more of the following constants:

**AT\_EMPTY\_PATH**

:   If *pathname* is an empty string, operate on the file referred to by
    *dirfd* (which may have been obtained from **open**(2) with
    **O\_PATH**, from **fsmount**(2) or from another **open\_tree**()).
    If *dirfd* is **AT\_FDCWD**, the call operates on the current
    working directory. In this case, *dirfd* can refer to any type of
    file, not just a directory. This flag is Linux-specific; define
    **\_GNU\_SOURCE** to obtain its definition.

**AT\_NO\_AUTOMOUNT**

:   Don\'t automount the terminal (\"basename\") component of *pathname*
    if it is a directory that is an automount point. This flag allows
    the automount point itself to be picked up or a mount cloned that is
    rooted on the automount point. The **AT\_NO\_AUTOMOUNT** flag has no
    effect if the mount point has already been mounted over. This flag
    is Linux-specific; define **\_GNU\_SOURCE** to obtain its
    definition.

**AT\_SYMLINK\_NOFOLLOW**

:   If *pathname* is a symbolic link, do not dereference it: instead
    pick up or clone a mount rooted on the link itself.

**OPEN\_TREE\_CLOEXEC**

:   Set the close-on-exec flag for the new file descriptor. This will
    cause the file descriptor to be closed automatically when a process
    exec\'s.

**OPEN\_TREE\_CLONE**

:   Rather than directly attaching the selected object to the file
    descriptor, clone the object, set the root of the new mount object
    to that point and attach the clone to the file descriptor.

**AT\_RECURSIVE**

:   This is only permitted in conjunction with OPEN\_TREE\_CLONE. It
    causes the entire mount subtree rooted at the selected spot to be
    cloned rather than just that one mount object.

EXAMPLE
=======

The **open\_tree**() function can be used like the following:

>     fd1 = open_tree(AT_FDCWD, "/mnt", 0);
>     fd2 = open_tree(fd1, "",
>                     AT_EMPTY_PATH | OPEN_TREE_CLONE | AT_RECURSIVE);
>     move_mount(fd2, "", AT_FDCWD, "/mnt2", MOVE_MOUNT_F_EMPTY_PATH);

This would attach the path point for \"/mnt\" to fd1, then it would copy
the entire subtree at the point referred to by fd1 and attach that to
fd2; lastly, it would attach the clone to \"/mnt2\".

RETURN VALUE
============

On success, the new file descriptor is returned. On error, -1 is
returned, and *errno* is set appropriately.

ERRORS
======

**EACCES**

:   Search permission is denied for one of the directories in the path
    prefix of *pathname*. (See also **path\_resolution**(7).)

**EBADF**

:   *dirfd* is not a valid open file descriptor.

**EFAULT**

:   *pathname* is NULL or *pathname* point to a location outside the
    process\'s accessible address space.

**EINVAL**

:   Reserved flag specified in *flags*.

**ELOOP**

:   Too many symbolic links encountered while traversing the pathname.

**ENAMETOOLONG**

:   *pathname* is too long.

**ENOENT**

:   A component of *pathname* does not exist, or *pathname* is an empty
    string and **AT\_EMPTY\_PATH** was not specified in *flags*.

**ENOMEM**

:   Out of memory (i.e., kernel memory).

**ENOTDIR**

:   A component of the path prefix of *pathname* is not a directory or
    *pathname* is relative and *dirfd* is a file descriptor referring to
    a file other than a directory.

VERSIONS
========

**open\_tree**() was added to Linux in kernel 4.18.

CONFORMING TO
=============

**open\_tree**() is Linux-specific.

NOTES
=====

Glibc does not (yet) provide a wrapper for the **open\_tree**() system
call; call it using **syscall**(2).

SEE ALSO
========

**fsmount**(2), **move\_mount**(2), **open**(2)

**Comment by asomers** [comment]
Do you know what problem these new syscalls are intended to solve?

**Comment by mahkoh** [comment]
The original motivation and an older version of the api is discussed [here](https://lwn.net/Articles/759499/).

I'm particularly interested in the ability to create anonymous mounts. That is, mounts that are not visible in the file system:

```c
#include <sys/syscall.h>
#include <linux/mount.h>
#include <unistd.h>
#include <fcntl.h>

int main(void) {
	int cfg = syscall(SYS_fsopen, "proc", 0);
	syscall(SYS_fsconfig, cfg, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
	int mnt = syscall(SYS_fsmount, cfg, 0, 0);
	int fd = openat(mnt, "cpuinfo", O_RDONLY);
	char buf[128];
	while (1) {
		int bytes = read(fd, buf, sizeof(buf));
		if (bytes <= 0) {
			break;
		}
		write(1, buf, bytes);
	}
}
```

**Comment by mahkoh** [comment]
Implemented [here](https://github.com/mahkoh/mount-api/blob/master/src/lib.rs).

**Comment by asomers** [comment]
The article only mentions that the new API relieves the requirement for all mount options to fit within 4096 bytes.  Is that the only benefit?

**Comment by mahkoh** [comment]
Thank you for taking the time to look into this.

Unfortunately, after trying to use nix in an application I'm working on, I found that I'm not really comfortable with the nix API. I've written my own bindings instead: https://github.com/mahkoh/uapi

Therefore I'm not interested into pushing for the inclusion of the new mount API into nix at this time. Of course you (or anyone else) can still use the nix-style code I posted above and include it in nix under the nix license.
